File: _A_arm/python/hw14/armController.py
import numpy as np
import armParamHW14 as P

class armController:
    def __init__(self):
        self.observer_state = np.array([
            [0.0],  # estimate of theta
            [0.0],  # estimate of theta_hat
            [0.0],  # estimate of disturbance
        ])
        self.tau_d1 = 0.0      # control torque, delayed 1 sample
        self.integrator = 0.0  # integrator
        self.error_d1 = 0.0    # error signal delayed by 1 sample
        self.K = P.K           # state feedback gain
        self.ki = P.ki         # Input gain
        self.L = P.L           # observer gain
        self.Ld = P.Ld
        self.L2 = P.L2
        self.A2 = P.A2         # system model
        self.B2 = P.B2
        self.C2 = P.C2
        self.limit = P.tau_max # Maxiumum force
        self.Ts = P.Ts         # sample rate of controller

    def update(self, theta_r, y_m):
        # update the observer and extract theta_hat
        x_hat, d_hat = self.update_observer(y_m)
        theta_hat = x_hat.item(0)

        # integrate error
        error = theta_r - theta_hat
        self.integrateError(error)

        # compute feedback linearizing torque tau_fl
        tau_fl = P.m * P.g * (P.ell / 2.0) * np.cos(theta_hat)

        # Compute the state feedback controller
        tau_tilde = -self.K @ x_hat \
                            - self.ki * self.integrator - d_hat

        # compute total torque
        tau = self.saturate(tau_fl + tau_tilde.item(0))
        self.tau_d1 = tau

        return tau, x_hat, d_hat

    def update_observer(self, y_m):
        # update the observer using RK4 integration
        F1 = self.observer_f(self.observer_state, y_m)
        F2 = self.observer_f(self.observer_state \
                             + self.Ts / 2 * F1, y_m)
        F3 = self.observer_f(self.observer_state \
                             + self.Ts / 2 * F2, y_m)
        F4 = self.observer_f(self.observer_state \
                             + self.Ts * F3, y_m)
        self.observer_state += self.Ts / 6 * \
                               (F1 + 2 * F2 + 2 * F3 + F4)
        x_hat = np.array([[self.observer_state.item(0)],
                          [self.observer_state.item(1)]])
        d_hat = self.observer_state.item(2)

        return x_hat, d_hat

    def observer_f(self, x_hat, y_m):
        # compute feedback linearizing torque tau_fl
        theta_hat = x_hat.item(0)
        tau_fl = P.m * P.g * (P.ell / 2.0) * np.cos(theta_hat)
        # xhatdot = A*xhat + B*(u-ue) + L(y-C*xhat)
        xhat_dot = self.A2 @ x_hat\
                   + self.B2 * (self.tau_d1 - tau_fl)\
                   + self.L2 * (y_m - self.C2 @ x_hat)

        return xhat_dot

    def integrateError(self, error):
        self.integrator = self.integrator \
                          + (self.Ts/2.0)*(error + self.error_d1)
        self.error_d1 = error

    def saturate(self,u):
        if abs(u) > self.limit:
            u = self.limit*np.sign(u)
        return u

File: _A_arm/matlab/hw14/armController.m
classdef armController < handle
    properties
        m
        ell
        g
        obsv_state
        tau_d1
        integrator
        error_d1
        K
        ki
        L
        Ld
        A
        B
        C
        limit
        Ts
        kr
    end
    methods
        %--------constructor--------------------
        function self = armController(P)
            % plant parameters known to controller
            self.m = P.m;
            self.ell = P.ell;
            self.g = P.g;
            % initialized object properties
            self.obsv_state = [0.0; 0.0; 0.0];
            self.tau_d1 = 0.0;
            self.integrator = 0.0;
            self.error_d1 = 0.0;
            self.K = P.K;
            self.ki = P.ki;
            self.A = P.A2;
            self.B  = P.B1;
            self.C  = P.C2;
            self.L  = P.L2;
            self.limit = P.tau_max;
            self.Ts = P.Ts;
            self.kr = 0;%P.kr;
        end
        function [tau, x_hat, d_hat] = update(self, theta_r, y)
            % update the observer and extract z_hat
            [x_hat, d_hat] = self.updateObserver(y);
            theta_hat = x_hat(1);

            % integrate error
            error = theta_r - theta_hat;
            self.integrateError(error);

            % compute feedback linearizing torque tau_fl
            tau_fl = self.m*self.g*(self.ell/2)*cos(theta_hat);

            % Compute the state feedback controller
            tau_tilde = -self.K * x_hat...
                - self.ki * self.integrator...
                - d_hat;

            % compute total torque
            tau = self.saturate( tau_fl + tau_tilde);
            self.tau_d1 = tau;  
        end
        function [xhat, dhat] = updateObserver(self, y)
            % update observer using RK4 integration
            F1 = self.obsv_f(self.obsv_state, y);
            F2 = self.obsv_f(self.obsv_state + self.Ts/2*F1, y);
            F3 = self.obsv_f(self.obsv_state + self.Ts/2*F2, y);
            F4 = self.obsv_f(self.obsv_state + self.Ts*F3, y);
            self.obsv_state = self.obsv_state ...
                + self.Ts/6 * (F1 + 2*F2 + 2*F3 + F4);
            xhat = self.obsv_state(1:2);
            dhat = self.obsv_state(3);
        end
        function x_hat_dot = obsv_f(self, obsv_state, y)
            % compute feedback linearizing torque tau_fl
            theta_hat = obsv_state(1);
            tau_fl = self.m*self.g*(self.ell/2)*cos(theta_hat);
            x_hat_dot = self.A * obsv_state...
                        + self.B * (self.tau_d1 - tau_fl)...
                        + self.L * (y - self.C * obsv_state);
        end
        function self = integrateError(self, error)
            self.integrator = self.integrator ...
                + (self.Ts/2.0)*(error+self.error_d1);
            self.error_d1 = error;
        end
        function out = saturate(self,u)
            if abs(u) > self.limit
                u = self.limit*sign(u);
            end
            out = u;
        end        
    end
end